io/src/lib.rs

```diff
@@ -0,0 +1,509 @@
+            }
+
+            impl ErrorKind {
+                fn description(&self) -> &'static str {
+                    match self {
+                        $(Self::$kind => stringify!($kind)),*
+                    }
+                }
+                #[cfg(feature = "std")]
+                fn to_std(self) -> std::io::ErrorKind {
+                    match self {
+                        $(Self::$kind => std::io::ErrorKind::$kind),*
+                    }
+                }
+                #[cfg(feature = "std")]
+                fn from_std(o: std::io::ErrorKind) -> ErrorKind {
+                    match o {
+                        $(std::io::ErrorKind::$kind => ErrorKind::$kind),*,
+                        _ => ErrorKind::Other
+                    }
+                }
+
+                #[cfg(all(feature = "core2", not(feature = "std")))]
+                fn to_core2(self) -> core2::io::ErrorKind {
+                    match self {
+                        $(Self::$kind => core2::io::ErrorKind::$kind),*
+                    }
+                }
+                #[cfg(all(feature = "core2", not(feature = "std")))]
+                fn from_core2(o: core2::io::ErrorKind) -> ErrorKind {
+                    match o {
+                        $(core2::io::ErrorKind::$kind => ErrorKind::$kind),*,
+                        _ => ErrorKind::Other
+                    }
+                }
+            }
+        }
+    }
+
+    define_errorkind!(
+        NotFound,
+        PermissionDenied,
+        ConnectionRefused,
+        ConnectionReset,
+        ConnectionAborted,
+        NotConnected,
+        AddrInUse,
+        AddrNotAvailable,
+        BrokenPipe,
+        AlreadyExists,
+        WouldBlock,
+        InvalidInput,
+        InvalidData,
+        TimedOut,
+        WriteZero,
+        Interrupted,
+        UnexpectedEof,
+        Other
+    );
+
+    pub type Result<T> = core::result::Result<T, Error>;
+
+    /// A generic trait describing an input stream. See [`std::io::Read`] for more info.
+    pub trait Read {
+        fn read(&mut self, buf: &mut [u8]) -> Result<usize>;
+        #[inline]
+        fn read_exact(&mut self, mut buf: &mut [u8]) -> Result<()> {
+            while !buf.is_empty() {
+                match self.read(buf) {
+                    Ok(0) =>
+                        return Err(ErrorKind::UnexpectedEof.into()),
+                    Ok(len) => {
+                        buf = &mut buf[len..];
+                    },
+                    Err(e) if e.kind() == ErrorKind::Interrupted => {},
+                    Err(e) => return Err(e),
+                }
+            }
+            Ok(())
+        }
+        #[inline]
+        fn take(&mut self, limit: u64) -> Take<Self> {
+            Take { reader: self, remaining: limit }
+        }
+        #[inline]
+        #[cfg(any(feature = "alloc", feature = "std"))]
+        fn read_to_end(&mut self, buf: &mut alloc::vec::Vec<u8>) -> Result<usize> {
```

Comment:
In 9a1c9e57cbefce13d1ee19b0b8f026a38530c893:

I am actually a little tempted to go in the opposite direction, changing behavior from stdlib to add a max allocation size and return an error if we over-read. This function has always made me a little nervous (same with `read_to_string`) since it's a memory-DoS vector.

In either case, can wait for a followup PR.

Reply:
Interesting, I didn't think of that.

Reply:
There's an argument to be made that `bitcoin`&co should never call this method so perhaps it'd make sense to remove it completely?

Reply:
At a single place, which I find hard to audit if it's actually DoS-free. (I guess it is because of how consensus decoding works; but why does PSBT use consensus decoding?)

The argument was to remove it as a lint so that we don't cause a DoS vector by accident.

Reply:
Everywhere that we read vectors we check the varint-encoded length and reject if it's too high.

AFAIK we don't use `read_to_end` or `read_to_string` anywhere in this library exactly because they are uncontrolled DoS vectors. It is true that you can simulate them to get a DoS anyway, but that's no reason for us to implement them.

Reply:
Exactly, we wrote a whole bunch of trait methods just to solve the DoS vector for consensus-encoded types, `read_to_end` is a missing piece.

Reply:
I know, I already checked. But it's actually reimplemented there so the method is not needed. And still I wonder if we have a DoS vector. I wasn't able to find out.